lib/chef/resource/lwrp_base.rb

```diff
@@ -35,32 +35,42 @@ class LWRPBase < Resource
 
       # Evaluates the LWRP resource file and instantiates a new Resource class.
       def self.build_from_file(cookbook_name, filename, run_context)
-        resource_class = nil
-        rname = filename_to_qualified_string(cookbook_name, filename)
-
-        class_name = convert_to_class_name(rname)
-        if Resource.const_defined?(class_name, false)
-          Chef::Log.info("#{class_name} light-weight resource is already initialized -- Skipping loading #{filename}!")
-          Chef::Log.debug("Overriding already defined LWRPs is not supported anymore starting with Chef 12.")
-          resource_class = Resource.const_get(class_name)
-        else
-          resource_class = Class.new(self)
+        resource_name = filename_to_qualified_string(cookbook_name, filename)
+
+        node = run_context ? run_context.node : Chef::Node.new
+        existing_class = Chef::ResourceResolver.new(node, resource_name).resolve
 
-          Chef::Resource.const_set(class_name, resource_class)
-          resource_class.resource_name = rname
-          resource_class.run_context = run_context
-          resource_class.class_from_file(filename)
+        if existing_class
+          Chef::Log.info("Skipping LWRP resource #{filename} from cookbook #{cookbook_name}.  #{existing_class} already defined.")
+          Chef::Log.debug("Overriding resources with LWRPs is not supported anymore starting with Chef 12.")
+          return existing_class
+        end
```

Comment:
`windows_package` and `windows_service` should be fine, because they will add `provides` and the `package` and `service` classes do not (they rely on the provider mapping).

The current logic is, if there are multiple `provides` the winner is the first in alphabetical order (unless there is an entry in the priority array).

Here's the summary of how it works right now (before this patch).  We try all these in order:

1. **Resource.provider:** If `resource.provider` is non-nil, we return that.
2. **Priority array:** If there is a priority array for this platform, we pick the first Provider that responds. Only the last-loaded priority array for this platform is used (so that cookbooks can override the system).
3. **Provides?:** If any descendant of `Provider` responds favorably to `provides?`, we return the first match *in alphabetical order by class name*.
4. **Platform mapping:** If the old platform mapping has an entry, we return that.  If there are multiple matches, we take the most specific (os first, then platform, then platform version).
5. **Chef::Provider::Blah:** If we find a `Chef::Provider` in `Chef::Provider::Blah`, we return that.

After the patch, behavior does not change except deprecation warnings.  However, after Chef X, `provides` will become required for `Chef::Resource` classes to participate in DSL.  When you add `provides` to a class, it is picked by step 3, and it now becomes possible (depending on alphabet) for the superclass to be picked where the subclass used to be picked, if you were relying on that implicit ordering.

The smallest suggestion I have to fix that is to modify step 3 so that any Resource or Provider with `provides :blah, <anything>` is preferred over `provides :blah` with no arguments or block.  I think there are further steps we can take to improve the selection overall, but I think we should probably take those steps separately.